packages/x-components/src/components/sliding-panel.vue

```diff
@@ -39,70 +39,111 @@
+  /* eslint-disable @typescript-eslint/unbound-method */
   export default class SlidingPanel extends Vue {
     /**
      * Scroll factor that will dictate how much the scroll moves when pressing a navigation button.
+     *
+     * @public
      */
     @Prop({ default: 0.7 })
-    protected scrollFactor!: number;
+    public scrollFactor!: number;
 
     /**
      * Would make the navigation buttons visible when they're needed or always hide them.
+     *
+     * @public
      */
     @Prop({ default: true })
-    protected showButtons!: boolean;
+    public showButtons!: boolean;
 
     /**
-     * MutationObserver that watches for changes inside the wrapping div to update
-     * the scroll position.
+     *
+     * @public
      */
-    protected scrollObserver = new MutationObserver(
-      // eslint-disable-next-line @typescript-eslint/unbound-method
-      this.restoreAndUpdateScroll
-    );
+    @Prop({ default: true })
+    public resetOnContentChange!: boolean;
 
     /**
      * Indicates if the scroll is at the start of the sliding panel.
+     *
+     * @internal
      */
     protected isScrollAtStart = true;
 
     /**
      * Indicates if the scroll is at the end of the sliding panel.
+     *
+     * @internal
      */
     protected isScrollAtEnd = true;
 
     /**
      * HTMLElement referencing the scroll of the component.
+     *
+     * @internal
      */
     public $refs!: {
       scrollContainer: HTMLElement;
     };
 
+    /**
+     * CSS classes to apply based on the scroll position.
+     *
+     * @returns The CSS classes to apply.
+     *
+     * @internal
+     */
+    protected get cssClasses(): VueCSSClasses {
+      return {
+        'x-sliding-panel--at-start': this.isScrollAtStart,
+        'x-sliding-panel--at-end': this.isScrollAtEnd
+      };
+    }
+
+    /**
+     * Initialises browser platform code:
+     * - Creates a mutation observer to detect content changes and reset scroll position.
+     * - Stores initial size and scroll position values.
+     *
+     * @internal
+     */
     mounted(): void {
-      this.scrollObserver.observe(this.$refs.scrollContainer, {
-        attributes: false,
-        childList: true,
-        subtree: true,
-        characterData: false
+      const resizeObserver = new ResizeObserver(this.debouncedUpdateScrollPosition);
+      resizeObserver.observe(this.$el);
+
+      const contentChangedObserver = new MutationObserver(this.restoreAndUpdateScroll);
+      this.$watch(
+        () => this.resetOnContentChange,
+        shouldReset => {
+          if (shouldReset) {
+            contentChangedObserver.observe(this.$refs.scrollContainer, {
+              attributes: false,
+              childList: true,
+              subtree: true,
+              characterData: false
+            });
+          } else {
+            contentChangedObserver.disconnect();
+          }
+        },
+        { immediate: true }
+      );
+      this.$on('hook:beforeDestroy', () => {
```

Comment:
I'm new to Vue so I'm not sure but should't we call $off when the component is destroyed to prevent memory leaks here?

Reply:
That's a good question. AFAIK Vue cleans this listeners itself. In the doc I found a quite similar example on what I'm doing here: https://vuejs.org/v2/guide/components-edge-cases.html#Programmatic-Event-Listeners. If you prefer we can change it by `$once`, but I think it will have the same effect, as the component can only be destroyed once.

To give you more confidence, I wrote a little test to check it:

```ts
it('does things', async () => {
  const destroyed = jest.fn();
  const wrapper = mount(
    {
      template: `<Test v-if="shouldRender"/>`,
      components: {
        Test: Vue.extend({
          template: `<div />`,
          mounted() {
            this.$on('hook:beforeDestroy', destroyed);
          }
        })
      },
      props: ['shouldRender']
    },
    {
      propsData: {
        shouldRender: false
      }
    }
  );
  await wrapper.setProps({ shouldRender: true });
  expect(destroyed).not.toHaveBeenCalled();
  await wrapper.setProps({ shouldRender: false });
  await wrapper.setProps({ shouldRender: true });
  expect(destroyed).toHaveBeenCalledTimes(1);
  await wrapper.setProps({ shouldRender: false });
  await wrapper.setProps({ shouldRender: true });
  expect(destroyed).toHaveBeenCalledTimes(2);
});
```

Reply:
Thanks for the explanation! I was looking for something like that in the docs but did not find it. IMO it would be better change it to $once in order to avoid that someone has the same doubt in the future. Besides, in the docs link you shared they are using $once, I think it is just more clear.